bigchaindb_driver/connection.py

```diff
@@ -25,36 +31,79 @@ def __init__(self, *, node_url, headers=None):
         if headers:
             self.session.headers.update(headers)
 
-    def request(self, method, timeout=None, *, path=None, json=None,
-                params=None, headers=None, **kwargs):
+        self._retries = 0
+        self.backoff_time = None
+
+    def request(self, method, *, path=None, json=None,
+                params=None, headers=None, timeout=None, **kwargs):
         """Performs an HTTP requests for the specified arguments.
 
+           If the node has backoff time attached, waits till the time is due.
+
+           Backoff time is adjusted after the request is made.
+
         Args:
             method (str): HTTP method (e.g.: ``'GET'``).
             path (str): API endpoint path (e.g.: ``'/transactions'``).
             json (dict): JSON data to send along with the request.
-            params (dict)): Dictionary of URL (query) parameters.
+            params (dict): Dictionary of URL (query) parameters.
             headers (dict): Optional headers to pass to the request.
+            timeout (int): Optional timeout in seconds.
             kwargs: Optional keyword arguments.
 
         """
-        url = self.node_url + path if path else self.node_url
-        response = self.session.request(
-            method=method,
-            timeout=timeout,
-            url=url,
-            json=json,
-            params=params,
-            headers=headers,
-            **kwargs
-        )
+        backoff_timedelta = self.get_backoff_timedelta()
+
+        if timeout is not None and timeout < backoff_timedelta:
+            raise TimeoutError
+
+        if backoff_timedelta > 0:
+            time.sleep(backoff_timedelta)
+
+        connExc = None
+        timeout = timeout if timeout is None else timeout - backoff_timedelta
+        try:
+            response = self._request(
+                method=method,
+                timeout=timeout,
+                url=self.node_url + path if path else self.node_url,
+                json=json,
+                params=params,
+                headers=headers,
+                **kwargs,
+            )
+        except ConnectionError as err:
+            connExc = err
+            raise err
+        finally:
+            self.update_backoff_time(success=connExc is None)
+        return response
+
+    def get_backoff_timedelta(self, timeout=None):
```

Comment:
We don't really use the `timeout` in here

Reply:
Good catch!